File: consumer/python/redis-consumer-1.py
from kafka import KafkaConsumer
import redis
import os
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

# Kafka configuration
kafkaHost = os.environ.get("KF_HOST", "localhost")
kafkaPort = os.environ.get("KF_PORT", "9092")
kafkaTopic = os.environ.get("KF_TOPIC", "stock-market1")

# Redis configuration
redis_host = "redis-server"
redis_port = 6379

# Create Redis client
r = redis.Redis(host=redis_host, port=redis_port, db=0, decode_responses=True)


# Function to insert data into Redis
def insert_data_to_redis(key, message):
    # Using LPUSH to insert data at the head of the list and LTRIM to limit the list to the last 100 items
    r.lpush(key, message)
    r.ltrim(key, 0, 99)


# Function to read last `n` items from Redis
def read_data_from_redis(key, n=100):
    # Using LRANGE to get the last `n` items from the list
    return r.lrange(key, 0, n - 1)


# Kafka consumer configuration
consumer = KafkaConsumer(
    kafkaTopic,
    group_id="stock-group1",
    bootstrap_servers=f"{kafkaHost}:{kafkaPort}",
)


def run():
    for msg in consumer:
        # Decode message to string
        message = msg.value.decode("utf-8")
        print(f"Received message: {message}")

        # Insert data to Redis
        insert_data_to_redis(kafkaTopic, message)

        # Read and print last 10 items from Redis
        last_messages = read_data_from_redis(kafkaTopic, 10)
        print(f"Last 10 items from Redis: {last_messages}")


if __name__ == "__main__":
    run()

File: consumer/node/redis-consumer-1.js
const { Kafka } = require("kafkajs");
const redis = require("redis");

require("dotenv").config();

const kafkaHost = process.env.KF_HOST || "localhost";
const kafkaPort = process.env.KF_PORT || "9092";
const kafkaTopic = process.env.KF_TOPIC || "stock-market1";

const client = redis.createClient({
  socket: {
    host: "redis-server",
    port: 6379,
  },
});

const kafka = new Kafka({
  clientId: "consumer1",
  brokers: [`${kafkaHost}:${kafkaPort}`],
});
const consumer = kafka.consumer({ groupId: "stock-group1" });

// Connect to Redis client
client.connect();

// Function to insert data into Redis
async function insertDataIntoRedis(topic, message) {
  await client.lPush(topic, message);
  await client.lTrim(topic, 0, 99);
}

// Function to read items from Redis
async function readItemsFromRedis(topic, amountOfData) {
  return await client.lRange(topic, 0, amountOfData - 1);
}

async function run() {
  await consumer.connect();
  await consumer.subscribe({ topic: kafkaTopic, fromBeginning: true });

  await consumer.run({
    eachMessage: async ({ topic, partition, message }) => {
      console.log(`[Consumer] Received message: ${message.value.toString()}`);
      try {
        await insertDataIntoRedis(topic, message.value.toString());
      } catch (error) {
        console.error("Error inserting data into Redis:", error);
      }

      try {
        const prices = await readItemsFromRedis(topic, 10);
        console.log("Last 10 items from Redis DB:", prices);
      } catch (error) {
        console.error("Error fetching items from Redis:", error);
      }
    },
  });
}

run().catch(console.error);
